### PR TITLE
shlink route created with necessary functionalities

### PR DESCRIPTION
**Pull request by godchiSymbionix**
## Summary by CodeRabbit

- **New Features**
	- Introduced an API endpoint for creating and retrieving short links.
	- Enhanced data integrity with automatic unique identifier generation for short links.
	- Added functionality for retrieving multiple and single short links based on user ID and link ID.

- **Bug Fixes**
	- Ensured unique constraints on identifiers to prevent duplicate entries.

- **Chores**
	- Simplified dependency management by removing specific version resolutions for packages.

### CODE REVIEW

File: src/domain/models/shlink.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
+import { BaseModel } from "./base-model";
+import z from 'zod';
+
+export class ShlinkModel extends BaseModel{
```
Comment (jacob-khoza-symb) [inline src/domain/models/shlink.ts:4]: Please rename to SHLinkModel

File: src/domain/dtos/shlink.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+
+export class CreateShlinkDto {
```
Comment (jacob-khoza-symb) [inline src/domain/dtos/shlink.ts:2]: please rename to CreateSHLinkDto

File: src/domain/dtos/shlink.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+    userId: string
+    passcodeFailuresRemaining: number
+    configPasscode?: string
+    configExp?: Date
+    active: boolean
+    managementToken: string
+}
+
+export class ShlinkDto extends CreateShlinkDto{
```
Comment (jacob-khoza-symb) [inline src/domain/dtos/shlink.ts:11]: Please rename SHLinkDto

File: src/domain/dtos/shlink.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+
+export class CreateShlinkDto {
+    userId: string
+    passcodeFailuresRemaining: number
+    configPasscode?: string
+    configExp?: Date
+    active: boolean
+    managementToken: string
```
Comment (jacob-khoza-symb) [inline src/domain/dtos/shlink.ts:8]: Please make this db generated

File: src/domain/dtos/shlink.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+
+export class CreateShlinkDto {
+    userId: string
+    passcodeFailuresRemaining: number
```
Comment (jacob-khoza-symb) [inline src/domain/dtos/shlink.ts:4]: make this db or use case generated

File: src/app/api/v1/share-link/route.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    catch(error){
+        return handleApiValidationError(error);
+    }
+}
+
+export async function GET(request: Request) {
+
+    try{
+        
+        const newShlink = await repo.findMany({})
```
Comment (jacob-khoza-symb) [inline src/app/api/v1/share-link/route.ts:28]: pass userId as a parameter and only return shlinks fo that user. Use a query parameter for the userId

File: src/usecases/shlinks/get-shlink.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+import { mapEntityToModel } from "@/mappers/shlink-mapper";
+
+export const getSHLinkUseCase = async (context: {repo: ISHLinkRepository}, data: {id: string}): Promise<SHLinkModel> => {
+    const entity = await context.repo.findById(data.id);
+
+    return mapEntityToModel(entity);
+}
+
+
+export const getSHLinkDeactivatedUseCase = async (context: {repo: ISHLinkRepository}, data: {id: string}): Promise<SHLinkModel> => {
```
Comment (jacob-khoza-symb) [inline src/usecases/shlinks/get-shlink.ts:12]: rename this to getDeactivatedSHLinksUseCase

File: src/usecases/shlinks/get-shlink.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    const entity = await context.repo.findById(data.id);
+    if(entity) entity.active = false;
+
+    return mapEntityToModel(entity);
+}
+
+export const getSHLinkActivatedUseCase = async (context: {repo: ISHLinkRepository}, data: {id: string}): Promise<SHLinkModel> => {
+    const entity = await context.repo.findById(data.id);
```
Comment (jacob-khoza-symb) [inline src/usecases/shlinks/get-shlink.ts:20]: see above comment but change active: true

File: src/usecases/shlinks/get-shlink.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+
+export const getSHLinkDeactivatedUseCase = async (context: {repo: ISHLinkRepository}, data: {id: string}): Promise<SHLinkModel> => {
+    const entity = await context.repo.findById(data.id);
+    if(entity) entity.active = false;
+
+    return mapEntityToModel(entity);
+}
+
+export const getSHLinkActivatedUseCase = async (context: {repo: ISHLinkRepository}, data: {id: string}): Promise<SHLinkModel> => {
```
Comment (jacob-khoza-symb) [inline src/usecases/shlinks/get-shlink.ts:19]: rename method to getActiveSHLinksUseCase

File: src/usecases/shlinks/get-shlink.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (godchiSymbionix) [reply src/usecases/shlinks/get-shlink.ts:12]: The usecase is meant to deactivate a shlink. I would rename it to deactivatedSHLinksUseCase

File: src/usecases/shlinks/add-shlink.test.ts [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        // Set up mock repository
+        mockRepo = {
+            insert: jest.fn(),
+            // other methods of ISHLinkRepository can be mocked here if needed
+        };
+
+        mockContext = { repo: mockRepo as ISHLinkRepository };
+
+        // Mock the data
+        mockShlinkModel = { /* your SHLinkModel data */ } as SHLinkModel;
```
Comment (jacob-khoza-symb) [inline src/usecases/shlinks/add-shlink.test.ts:31]: please provide test values here

File: src/usecases/shlinks/get-shlink.ts [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+
+export const getSHLinkUseCase = async (context: {repo: ISHLinkRepository}, data: {id: string}): Promise<SHLinkModel> => {
+    const entity = await context.repo.findById(data.id);
+
+    return mapEntityToModel(entity);
+}
+
+
+export const deactivatedSHLinksUseCase = async (context: {repo: ISHLinkRepository}, data: {id: string}): Promise<SHLinkModel> => {
+    const entity = await context.repo.findById(data.id);
```
Comment (jacob-khoza-symb) [inline src/usecases/shlinks/get-shlink.ts:13]: is this saved to the data base?

File: src/domain/dtos/shlink.ts (hunk 4 above)
Comment (godchiSymbionix) [reply src/domain/dtos/shlink.ts:8]: db generated for active and managementToken right?

File: src/domain/dtos/shlink.ts (hunk 5 above)
Comment (godchiSymbionix) [reply src/domain/dtos/shlink.ts:4]: I think userId should come from the client side and passcodeFailuresRemaining is already defaulted to 5

File: src/app/api/v1/share-link/[id]/deactivate/route.ts [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+import { SHLinkPrismaRepository } from "@/infrastructure/repositories/prisma/shlink-repository";
+import { mapDtoToModel, mapEntityToModel, mapModelToDto } from "@/mappers/shlink-mapper";
+import { deactivatedSHLinksUseCase } from "@/usecases/shlinks/get-shlink";
+import { NextResponse } from "next/server";
+import { NOT_FOUND } from "@/app/constants/http-constants";
+
+const repo = new SHLinkPrismaRepository(prisma);
+
+  export async function GET(request: Request, { params }: { params: { id: string } }) {
+    console.log(params.id)
```
Comment (jacob-khoza-symb) [inline src/app/api/v1/share-link/[id]/deactivate/route.ts:13]: remove debug code

**Review by jacob-khoza-symb — APPROVED** [review]

**Pull request merged**